### ISSUE TITLE
Data race caught in clientapi/routing/register.go

### ISSUE THREAD

**Issue by Cnly** [matrix-org/dendrite#780]
Running SyTest with the race detector enabled, I was able to catch one data race:

```
==================
WARNING: DATA RACE
Read at 0x00c000172060 by goroutine 69:
  runtime.mapaccess2_faststr()
      /usr/local/go/src/runtime/map_faststr.go:107 +0x0
  github.com/matrix-org/dendrite/clientapi/routing.(*sessionsDict).AddCompletedStage()
      /home/user/git/dendrite/clientapi/routing/register.go:79 +0x8d
  github.com/matrix-org/dendrite/clientapi/routing.handleRegistrationFlow()
      /home/user/git/dendrite/clientapi/routing/register.go:576 +0x22c
  github.com/matrix-org/dendrite/clientapi/routing.Register()
      /home/user/git/dendrite/clientapi/routing/register.go:497 +0x826
  . . . (omitted)

Previous write at 0x00c000172060 by goroutine 83:
  runtime.mapassign_faststr()
      /usr/local/go/src/runtime/map_faststr.go:202 +0x0
  github.com/matrix-org/dendrite/clientapi/routing.(*sessionsDict).AddCompletedStage()
      /home/user/git/dendrite/clientapi/routing/register.go:94 +0x26f
  github.com/matrix-org/dendrite/clientapi/routing.handleRegistrationFlow()
      /home/user/git/dendrite/clientapi/routing/register.go:576 +0x22c
  github.com/matrix-org/dendrite/clientapi/routing.Register()
      /home/user/git/dendrite/clientapi/routing/register.go:497 +0x826
  . . . (omitted)
```

In some of my parallel test runs, this warning is followed by a panic (which may be the one we've seen earlier on CircleCI but was unable to trace since the log was truncated):

```
fatal error: concurrent map read and map write

goroutine 73735 [running]:
runtime.throw(0xf1143b, 0x21)
        /usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc000018e78 sp=0xc000018e48 pc=0x456b72
runtime.mapaccess2_faststr(0xe182c0, 0xc000182060, 0xc00033c040, 0x18, 0x16374a0, 0x0)
        /usr/local/go/src/runtime/map_faststr.go:116 +0x4e6 fp=0xc000018ee8 sp=0xc000018e78 pc=0x43bb96
github.com/matrix-org/dendrite/clientapi/routing.sessionsDict.GetCompletedStages(...)
        /home/user/git/dendrite/clientapi/routing/register.go:79
github.com/matrix-org/dendrite/clientapi/routing.(*sessionsDict).AddCompletedStage(0xc000010030, 0xc00033c040, 0x18, 0xf02cfe, 0xd)
        /home/user/git/dendrite/clientapi/routing/register.go:94 +0x178 fp=0xc000018f90 sp=0xc000018ee8 pc=0xc3adf8
github.com/matrix-org/dendrite/clientapi/routing.handleRegistrationFlow(0xc00079bd00, 0xc000052030, 0xe, 0xc00033c000, 0x18, 0x0, 0xc000052060, 0xd, 0x0, 0x0, ...)
        /home/user/git/dendrite/clientapi/routing/register.go:576 +0x22d fp=0xc0000192e8 sp=0xc000018f90 pc=0xc3f81d
github.com/matrix-org/dendrite/clientapi/routing.Register(0xc00079bd00, 0xc000294000, 0xc0000c48c0, 0xc0000d2d80, 0x0, 0x0, 0xdec8e0, 0xc0002d2bf0)
        /home/user/git/dendrite/clientapi/routing/register.go:497 +0x827 fp=0xc000019668 sp=0xc0000192e8 pc=0xc3eef7
```

Dendrite revision tested against: [`94ea325c`](https://github.com/matrix-org/dendrite/tree/94ea325c93d03362795ce85c49008797d37d36dd); register.go is [here](https://github.com/matrix-org/dendrite/blob/94ea325c93d03362795ce85c49008797d37d36dd/clientapi/routing/register.go#L79).

There didn't seem to be other data races during my tests.

**Comment by Cnly** [comment]
@anoadragon453 In https://github.com/matrix-org/dendrite/pull/405#discussion_r305693944 you said that you'll be replacing `func (d sessionsDict) GetCompletedStages` with `func AddCompletedSessionStage`. So should we fix this race in the new function in #405, instead of in the to-be-removed function?

**Comment by anoadragon453** [comment]
@Cnly Yes we should. Hopefully we can get #405 in asap.
